tpp-rt/PerfRunnerUtils.cpp

```diff
@@ -29,25 +33,73 @@ int64_t _mlir_ciface_timer_alloc() {
 
 void _mlir_ciface_timer_start(int64_t acc) {
   assert(acc >= 0 && (int64_t)timerResults.size() > acc && "Invalid timer ID");
-  auto& perfResults = timerResults[acc];
+  auto &perfResults = timerResults[acc];
   perfResults.startTimer();
 }
 
 void _mlir_ciface_timer_stop(int64_t acc) {
   assert(acc >= 0 && (int64_t)timerResults.size() > acc && "Invalid timer ID");
-  auto& perfResults = timerResults[acc];
+  auto &perfResults = timerResults[acc];
   perfResults.stopTimer();
 }
 
 double _mlir_ciface_timer_average(int64_t acc) {
   assert(acc >= 0 && (int64_t)timerResults.size() > acc && "Invalid timer ID");
-  auto& perfResults = timerResults[acc];
+  auto &perfResults = timerResults[acc];
   return perfResults.getMean();
 }
 
 double _mlir_ciface_timer_deviation(int64_t acc) {
   assert(acc >= 0 && (int64_t)timerResults.size() > acc && "Invalid timer ID");
-  auto& perfResults = timerResults[acc];
+  auto &perfResults = timerResults[acc];
   return perfResults.getStdev();
 }
 
+//===----------------------------------------------------------------------===//
+// Perf dialect utils
+//===----------------------------------------------------------------------===//
+
+// Return the current timestamp.
+int64_t _mlir_ciface_perf_start_timer() {
+  auto timestamp = std::chrono::high_resolution_clock::now();
+  return timestamp.time_since_epoch().count();
+}
+
+// Compute time delta between the starting time and now.
+double _mlir_ciface_perf_stop_timer(int64_t startTimestamp) {
+  auto stop = std::chrono::high_resolution_clock::now();
+  std::chrono::system_clock::time_point start{
+      std::chrono::system_clock::duration{startTimestamp}};
+  return std::chrono::duration_cast<std::chrono::duration<double>>(stop - start)
+      .count();
+}
+
+// A generic sink function.
+// Its aim is to ensure that the passed data and its producers cannot be
+// optimized away such that the time measured by a benchmark loop correctly
+// represents the full workload.
+static void __attribute__((optnone)) perf_sink(void *data) { (void)data; }
+
+/*
+  Perf dialect runtime bindings for common perf.sink op argument types.
+*/
+void _mlir_ciface_perf_sink_memref_i8(UnrankedMemRefType<int8_t> *val) {
```

Comment:
I don't think we need runtime support for this. Declaring the functions at the IR level and marking as opt-none / no-inline may be enough to get all compile time types needed.

Reply:
I wasn't sure how to make it work directly in MLIR before but I finally figured it out. It's much cleaner now.